controllers/mirroring/mirroring_controller.go

```diff
@@ -0,0 +1,478 @@
+	}
+	defer ocsClient.Close()
+
+	if err = r.reconcileRbdMirror(clientMappingConfig, disableMirroring); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if res, err := r.reconcileBlockPoolMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		disableMirroring,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if res, err := r.reconcileRadosNamespaceMirroring(
+		ocsClient,
+		clientMappingConfig,
+		storageClusterPeer,
+		disableMirroring,
+	); err != nil || !res.IsZero() {
+		return res, err
+	}
+
+	if disableMirroring {
+		if controllerutil.RemoveFinalizer(clientMappingConfig, mirroringFinalizer) {
+			r.log.Info("removing finalizer from ConfigMap.")
+			if err := r.update(clientMappingConfig); err != nil {
+				r.log.Info("Failed to remove finalizer from ConfigMap")
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from ConfigMap: %v", err)
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MirroringReconciler) reconcileRbdMirror(clientMappingConfig *corev1.ConfigMap, disableMirroring bool) error {
+	rbdMirror := &rookCephv1.CephRBDMirror{}
+	rbdMirror.Name = util.CephRBDMirrorName
+	rbdMirror.Namespace = clientMappingConfig.Namespace
+
+	storageConsumers := &ocsv1alpha1.StorageConsumerList{}
+	if err := r.list(
+		storageConsumers,
+		client.MatchingFields{util.AnnotationIndexName: util.RequestMaintenanceModeAnnotation},
+	); err != nil {
+		return fmt.Errorf("failed to list StorageConsumer(s): %v", err)
+	}
+
+	disableRBDMirror := disableMirroring || len(storageConsumers.Items) >= 1
+
+	if !disableRBDMirror {
+		_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, rbdMirror, func() error {
+			if err := r.own(clientMappingConfig, rbdMirror); err != nil {
+				return err
+			}
+			rbdMirror.Spec.Count = 1
+			return nil
+		})
+		if err != nil {
+			r.log.Error(err, "Failed to create/update the CephRBDMirror", "CephRBDMirror", rbdMirror)
+			return err
+		}
+	} else {
+		if err := r.delete(rbdMirror); err != nil {
+			return fmt.Errorf("failed to delete CephRBDMirror: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *MirroringReconciler) reconcileBlockPoolMirroring(
+	ocsClient *providerClient.OCSProviderClient,
+	clientMappingConfig *corev1.ConfigMap,
+	storageClusterPeer *ocsv1.StorageClusterPeer,
+	disableMirroring bool,
+) (ctrl.Result, error) {
+
+	cephBlockPoolsList := &rookCephv1.CephBlockPoolList{}
+	if err := r.list(
+		cephBlockPoolsList,
+		client.InNamespace(clientMappingConfig.Namespace),
+	); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list cephBlockPools: %v", err)
+	}
+
+	blockPoolByName := map[string]*rookCephv1.CephBlockPool{}
+	for i := range cephBlockPoolsList.Items {
+		cephBlockPool := &cephBlockPoolsList.Items[i]
+		if _, ok := cephBlockPool.GetLabels()[util.ForbidMirroringLabel]; !ok {
+			continue
+		}
+		blockPoolByName[cephBlockPoolsList.Items[i].Name] = &cephBlockPoolsList.Items[i]
```

Comment:
You have a local var, use it.
```suggestion
		blockPoolByName[cephBlockPool.Name] = cephBlockPool
```